src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -322,49 +322,73 @@ Klass* SystemDictionary::resolve_array_class_or_null(Symbol* class_name,
-//      loadClassFromxxx or defineClass via parseClassFile Super ...
-//      3.1 resolve_super_or_fail creates placeholder: T1, Super (super Base)
-//      3.3 resolve_instance_class_or_null Base, finds placeholder for Base
-//      3.4 calls resolve_super_or_fail Base
-//      3.5 finds T1,Base -> throws class circularity
-//OR 4. If T2 tries to resolve Super via defineClass Super ...
-//      4.1 resolve_super_or_fail creates placeholder: T2, Super (super Base)
-//      4.2 resolve_instance_class_or_null Base, finds placeholder for Base (super Super)
-//      4.3 calls resolve_super_or_fail Super in parallel on own thread T2
-//      4.4 finds T2, Super -> throws class circularity
-// Be careful when modifying this code: once you have run
-// placeholders()->find_and_add(PlaceholderTable::LOAD_SUPER),
-// you need to find_and_remove it before returning.
-// So be careful to not exit with a CHECK_ macro between these calls.
+//      If another thread is trying to resolve the class, it must do
+//      super-class checks on its own thread to catch class circularity and
+//      to avoid deadlock.
+//
+// Take the case: A extends B extends A
+//
+// Single Threaded:
+//   1. forName/ConstantPool reference for A: calls resolve_or_fail for A
+//      Bootstrap and non-parallelCapable create placeholder for LOAD_INSTANCE T1 A
+//      loadClass A (parallelCapable loader locks A)
+//   2. defineClass A -> calls resolve_from_stream -> calls resolve_super_or_fail for B
+//      resolve_super_or_fail creates placeholder for LOAD_SUPER T1 A, superclass B
+//   3. resolve_or_fail for B
+//      Bootstrap and non-parallelCapable create placeholder LOAD_INSTANCE T1 B
+//      loadClass B (parallelCapable loader locks B)
+//   4. defineClass B -> calls resolve_from_stream -> calls resolve_super_or_fail for A
+//      resolve_super_or_fail creates placeholder for LOAD_SUPER T1 B, superclass A
+//   5. resolve_or_fail for A
+//      if Bootstrap and non-parallelCapable class loader:
+//         Finds placeholder for LOAD_INSTANCE T1 A -> throws ClassCircularityError
+//      else
+//         loadClass A
+//   6. defineClass A -> calls resolve_from_stream -> calls resolve_super_or_fail for B
+//      Finds placeholder LOAD_SUPER T1 A superclass B -> throws ClassCircularityError
+//
+// Class circularity checking for cases where classloading is delegated to
+// different threads and the classloader lock is released, is described below:
+//
+// Step 3: If loadClass B above stalls due to ClassLoader.wait() call, or bootstrap class
+// loading losing the race to acquire the SystemDictionary_lock, the second thread will attempt
+// to load B to detect CCE, then waits until the first thread completes loading 4-6 and then
+// resumes steps 4-6.
+//
+// For ParallelCapable class loaders, will call handle_parallel_super_load to attempt to load
+// the super class directly, then resume steps 4-6.
+//
+//
```

Comment:
resolve_super_or_fail had this comment that I found difficult to understand and work through, which I did at one point, so I wanted to replace it with one that seemed less so.  It occurs to me that this is probably unhelpful also.   Maybe the whole comment from lines 351 to 381 should be replaced with a sentence like:

//resolve_super_or_fail adds a LOAD_SUPER placeholder to the placeholder table before calling
// resolve_instance_class_or_null.  ClassCircularityError is detected when a LOAD_SUPER or LOAD_INSTANCE
// placeholder for the same thread, class, classloader is found.

I'm unlikely to work through the new comment example again myself.

Reply:
Since nobody wants to read this example in this comment and neither do I, I'm removing it.

Reply:
I guess it's OK to remove this comment, since most of us who have looked at this code in the past few years haven't found this comment to be helpful. Maybe you can add a comment about `-Xlog:class+load+placeholders=debug` instead?

If we do want to improve the comment, maybe we can replacing Base->Super with something like Car->Vehicle. That way it's clear which is the supertype.

Reply:
Thanks Ioi.